Update controlling-deployments-with-barriers-resource-constraints-and… (#4328)
harness/developer-hub: docs/continuous-delivery/manage-deployments/controlling-deployments-with-barriers-resource-constraints-and-queue-steps.md

### Resource Constraints

> Resource Constraints protect resource capacity limits by preventing simultaneous deployments to the same Service + Infrastructure combination. The Service + Infrastructure combination acts as a fixed key called Infrastructure Key. The Infrastructure key (the unique key used to restrict concurrent deployments) gets formed with service + environment + connector + InfrastructureID. This key by default is scoped at the account level. This means if the infrastructure is being reused between different projects when the service + infra + environment + connector have the same IDs it can result in queuing of the infrastructure. Your Pipelines will build up in a waiting state.

Reviewer: ```suggestion
Resource Constraints protect resource capacity limits by preventing simultaneous deployments to the same Service + Infrastructure combination. The Harness service + infrastructure combination acts as a fixed key called the Infrastructure Key. The Infrastructure Key (the unique key used to restrict concurrent deployments) is formed using a combination of Harness service Id + environment Id + connector Id + infrastructure Id. 

By default, this key is scoped at the Harness account level. This means if the infrastructure is reused in different Harness projects with the same combination of service Id + environment Id + connector Id + infrastructure Id, it can result in queuing of the infrastructure. Consequently, your pipelines will build up in a waiting state.
```

Resource Constraints are added to every Stage by default, but it can be disabled in a Stage's **Infrastructure** settings by enabling the **Allow simultaneous deployments on the same infrastructure** option.

See [Pipeline Resource Constraints](./deployment-resource-constraints.md).

The automatic **Resource Constraints** setting does not apply to [Custom stages](/docs/platform/pipelines/add-a-stage.md#add-a-custom-stage). **Resource Constraints** apply to a combination of Service + Infrastructure, and Custom Stages have no Services or Infrastructures. You can use Barriers and Queue steps in any stage types.

> #### Resource Constraint Scoping at Project Level

Reviewer: ```suggestion
### Resource constraint scoping at the project level
```

:::note

Currently, the dynamic provisioning documented in this topic is behind the feature flag `CDS_PROJECT_SCOPED_RESOURCE_CONSTRAINT_QUEUE`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

> By default resource constraints are scoped at a account level. To recap, this means if the infrastructure is being reused between different projects when the service + infra + namespace have the same IDs it can result in queuing of the infrastructure. 

Reviewer: ```suggestion
By default resource constraints are scoped at a account level. This means if the infrastructure is reused in different Harness projects with the same combination of service Id + environment Id + connector Id + infrastructure Id, it can result in queuing of the infrastructure. Consequently, your pipelines will build up in a waiting state. 
```

> With this feature, users can now scope the resource constraint down to the individual project level. This ensures that the resource constraint is scoped to a particular project and doesn't impact other projects using the exact same service, environment, infrastructure, and connectors. The Infrastructure key (the unique key used to restrict concurrent deployments) now gets formed with accountID + OrgID + ProjectID + service + environment + connector + infrastructureID. This ensures that there is uniqueness across the various projects and it won't build a queue because of other project teams deploying into the same infrastructure.

Reviewer: ```suggestion
With Resource constraint scoping at the project level, users can now scope the resource constraint down to the individual project. 

This scoping ensures that the resource constraint is scoped to a particular project and doesn't impact other projects using the exact same combination of service Id + environment Id + connector Id + infrastructure Id. 

With project-level scoping, the Infrastructure key (the unique key used to restrict concurrent deployments) is now formed with the Harness account Id + org Id + project Id + service Id + environment Id + connector Id + infrastructure Id. 

This scoping ensures that there are unique resources across the various projects and prevents a queue because of other project teams deploying into the same infrastructure.
```
